File: navi/src/main/java/org/kexie/android/dng/navi/viewmodel/QueryViewModel.java
package org.kexie.android.dng.navi.viewmodel;

import android.app.Application;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.text.TextUtils;

import com.amap.api.services.core.AMapException;
import com.amap.api.services.core.LatLonPoint;
import com.amap.api.services.core.PoiItem;
import com.amap.api.services.help.Inputtips;
import com.amap.api.services.help.InputtipsQuery;
import com.amap.api.services.poisearch.PoiSearch;
import com.amap.api.services.route.RouteSearch;
import com.orhanobut.logger.Logger;

import org.kexie.android.common.databinding.GenericQuickAdapter;
import org.kexie.android.dng.navi.model.Point;
import org.kexie.android.dng.navi.model.Query;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.WorkerThread;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.subjects.PublishSubject;
import java8.util.stream.Collectors;
import java8.util.stream.StreamSupport;
import mapper.Request;

public class QueryViewModel extends AndroidViewModel
{
    public static final String DEBUG_TEXT = "火车站";

    private static final String CITY = "西安";

    private final RouteSearch routeSearch;

    private final Executor singleTask = Executors.newSingleThreadExecutor();

    private final MutableLiveData<String> queryText = new MutableLiveData<>();

    private final MutableLiveData<List<Request>> routes = new MutableLiveData<>();

    private final PublishSubject<String> onLoading = PublishSubject.create();

    private final PublishSubject<String> onErrorMessage = PublishSubject.create();

    private final PublishSubject<String> onSuccessMessage = PublishSubject.create();

    private final List<String> poiIds = new ArrayList<>();

    private GenericQuickAdapter<String> adapter;

    public void bindAdapter(GenericQuickAdapter<String> adapter)
    {
        this.adapter = adapter;
    }

    public QueryViewModel(@NonNull Application application)
    {
        super(application);
        routeSearch = new RouteSearch(application);
    }

    public LiveData<List<Request>> getRoutes()
    {
        return routes;
    }

    @MainThread
    public void tipQueryBy(String text)
    {
        singleTask.execute(() -> {
            Handler handler = new Handler(Looper.getMainLooper());
            handler.post(() -> {
                adapter.getData().clear();
                adapter.notifyDataSetChanged();
                poiIds.clear();
            });
            InputtipsQuery inputtipsQuery = new InputtipsQuery(text, CITY);
            Inputtips inputtips = new Inputtips(getApplication(), inputtipsQuery);
            try
            {
                StreamSupport.stream(inputtips.requestInputtips())
                        .filter(tip -> !TextUtils.isEmpty(tip.getPoiID()))
                        .forEach(tip -> handler.post(() -> {
                            adapter.addData(tip.getName());
                            poiIds.add(tip.getPoiID());
                        }));
            } catch (Exception e)
            {
                e.printStackTrace();
                onErrorMessage.onNext("输入提示查询失败,请检查网络连接");
            }
        });
    }

    @MainThread
    public void routeQueryBy(String tip)
    {
        int index = adapter.getData().indexOf(tip);
        String poiId = poiIds.get(index);
        if (poiId != null)
        {
            singleTask.execute(() -> {
                Point point = getTipPoint(tip, poiId);
                Query query = new Query.Builder()
                        .from(Point.form(108.947167,34.309997))
                        .to(point)
                        .build();
                routeQuery(query);
            });
        }
    }

    public Observable<String> getOnSuccessMessage()
    {
        return onSuccessMessage.observeOn(AndroidSchedulers.mainThread());
    }

    public Observable<String> getOnErrorMessage()
    {
        return onErrorMessage.observeOn(AndroidSchedulers.mainThread());
    }

    public Observable<String> getOnLoading()
    {
        return onLoading;
    }

    @WorkerThread
    private void routeQuery(Query query)
    {
        List<List<LatLonPoint>> avoids = query.avoids == null
                ? Collections.emptyList()
                : StreamSupport.stream(query.avoids)
                .map(points -> StreamSupport.stream(points)
                        .map(point -> point.unBox(LatLonPoint.class))
                        .collect(Collectors.toList()))
                .collect(Collectors.toList());

        LatLonPoint form = query.from == null ? null : query.from.unBox(LatLonPoint.class);


        LatLonPoint to = query.to == null ? null : query.to.unBox(LatLonPoint.class);

        List<LatLonPoint> ways = query.ways == null
                ? Collections.emptyList()
                : StreamSupport.stream(query.ways)
                .map(point -> point.unBox(LatLonPoint.class))
                .collect(Collectors.toList());

        RouteSearch.DriveRouteQuery driveRouteQuery
                = new RouteSearch.DriveRouteQuery(
                new RouteSearch.FromAndTo(form, to),
                query.mode,
                ways,
                avoids, "");
        try
        {
            List<Request> requests = StreamSupport.stream(routeSearch
                    .calculateDriveRoute(driveRouteQuery)
                    .getPaths())
                    .map(path -> {
                        Bundle bundle = new Bundle();
                        bundle.putParcelable("path", path);
                        bundle.putParcelable("from", query.from);
                        bundle.putParcelable("to", query.to);
                        return new Request.Builder()
                                .bundle(bundle)
                                .uri("dng/navi/route")
                                .build();
                    }).collect(Collectors.toList());
            Logger.d("route size "+requests.size());
            routes.postValue(requests);
        } catch (Exception e)
        {
            e.printStackTrace();
            onErrorMessage.onNext("路径规划失败,请检查网络连接");
            routes.postValue(null);
        }
    }

    @MainThread
    private Point getTipPoint(String tip, String poiId)
    {
        PoiSearch.Query query = new PoiSearch.Query(tip, "");
        query.setDistanceSort(false);
        query.requireSubPois(true);
        PoiSearch poiSearch = new PoiSearch(getApplication(), query);
        try
        {
            PoiItem item = poiSearch.searchPOIId(poiId);
            LatLonPoint latLonPoint
                    = ((latLonPoint = item.getEnter()) != null)
                    ? (latLonPoint)
                    : ((latLonPoint = item.getExit()) != null
                    ? latLonPoint
                    : item.getLatLonPoint());
            return Point.box(latLonPoint);
        } catch (AMapException e)
        {
            onErrorMessage.onNext("目标地点信息读取失败,请检查网络连接");
            e.printStackTrace();
            return null;
        }
    }

}

File: navi/src/main/java/org/kexie/android/dng/navi/viewmodel/QueryViewModel.kt
package org.kexie.android.dng.navi.viewmodel

import android.app.Application
import android.os.HandlerThread
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.MutableLiveData
import com.amap.api.maps.AMapException
import com.amap.api.navi.model.NaviLatLng
import com.amap.api.navi.model.NaviPath
import com.amap.api.services.core.PoiItem
import com.amap.api.services.poisearch.PoiSearch
import com.blankj.utilcode.util.NetworkUtils
import com.orhanobut.logger.Logger
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.exceptions.Exceptions
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.PublishSubject
import org.kexie.android.dng.navi.model.Point
import org.kexie.android.dng.navi.model.Query
import org.kexie.android.dng.navi.viewmodel.entity.GuideInfo
import org.kexie.android.dng.navi.viewmodel.entity.InputTip
import org.kexie.android.dng.navi.viewmodel.entity.RouteInfo
import org.kexie.android.dng.navi.widget.AMapCompat
import org.kexie.android.dng.navi.widget.NaviCallback
import java.text.DecimalFormat
import java.util.*
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock
typealias NaviController = com.amap.api.navi.AMapNavi;

class QueryViewModel(application: Application,private val navi:NaviController)
    : AndroidViewModel(application) {

    private val worker = HandlerThread(toString())
            .apply {
                start()
            }

    var networkTest: Boolean = false
        private set

    val routes = MutableLiveData<Map<Int, RouteInfo>>()

    val currentSelect = MutableLiveData<Int>()

    val isLoading = MutableLiveData<Boolean>()

    val onError = PublishSubject.create<String>()

    val onSuccess = PublishSubject.create<String>()

    init {
        ping()
    }

    fun query(query: Query) {
        isLoading.value = true
        query0(Observable.just(query))
    }

    fun select(id: Int) {
        currentSelect.value = id;
        if (id != NO_SELECT) {
            navi.selectRouteId(id)
        }
    }

    fun query(inputTip: InputTip) {

        isLoading.value = true

        val query = Observable.just(inputTip)
                .observeOn(Schedulers.io())
                .map {
                    val query = PoiSearch.Query(it.text, "")
                            .apply {
                                isDistanceSort = false
                                requireSubPois(true)
                            }
                    val search = PoiSearch(getApplication(), query)
                    try {
                        val item = search.searchPOIId(it.poiId)
                        getPoiPoint(item)
                    } catch (e: AMapException) {
                        throw Exceptions.propagate(e)
                    }
                }.map {
                    Query.Builder()
                            .to(it)
                            .mode(10)
                            .build()!!
                }

        query0(query)

    }

    private fun query0(query: Observable<Query>) {

        query.observeOn(AndroidSchedulers.from(worker.looper))
                .map {
                    val ids = getRouteIds(it)
                    if (ids.isEmpty())
                        emptyMap()
                    else
                        ids.asSequence().map { x ->
                            x to getRouteInfo(x)
                        }.toMap()
                }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Observer<Map<Int, RouteInfo>> {
                    override fun onSubscribe(d: Disposable) {

                    }

                    override fun onNext(t: Map<Int, RouteInfo>) {
                        routes.value = t
                        if (!t.isEmpty()) {
                            select(t.keys.first())
                            onSuccess.onNext("路径规划成功")
                        }
                    }

                    override fun onError(e: Throwable) {
                        onError.onNext("路径规划失败,请检查网络连接")
                    }

                    override fun onComplete() {
                        isLoading.value = false
                    }
                })
    }

    private fun getRouteIds(query: Query): Array<Int> {
        return try {
            val lock = ReentrantLock()

            val condition = lock.newCondition()

            var result: IntArray? = null

            navi.addAMapNaviListener(
                    object : NaviCallback() {
                        @Suppress("OverridingDeprecatedMember")
                        override fun onCalculateRouteFailure(i: Int) {
                            lock.withLock {
                                navi.removeAMapNaviListener(this)
                                Logger.d("error code $i")
                                condition.signalAll()
                            }
                        }

                        @Suppress("OverridingDeprecatedMember")
                        override fun onCalculateRouteSuccess(ints: IntArray) {
                            lock.withLock {
                                navi.removeAMapNaviListener(this)
                                result = ints
                                condition.signalAll()
                            }
                        }
                    })

            lock.withLock {

                val to = if (query.to == null)
                    emptyList()
                else
                    listOf(query.to.unBox(NaviLatLng::class.java))
                val ways = if (query.ways.isNullOrEmpty())
                    emptyList()
                else
                    query.ways.map { p -> p.unBox(NaviLatLng::class.java) }
                            .toList()

                navi.calculateDriveRoute(to, ways, query.mode)

                condition.await()

            }

            result?.toTypedArray() ?: emptyArray()
        } catch (e: Exception) {
            e.printStackTrace()
            throw Exceptions.propagate(e)
        }
    }

    override fun onCleared() {
        navi.destroy()
        worker.quit()
    }

    private fun getRouteInfo(id: Int): RouteInfo {
        val path = AMapCompat.getAllNaviPath(navi)[id]!!
        return RouteInfo.Builder()
                .length(getPathLength(path.allLength))
                .time(getPathTime(path.allTime))
                .name(path.amapNaviPath.labels)
                .path(path.amapNaviPath)
                .guideInfos(getGuideInfo(path))
                .build()
    }

    @Suppress("MissingPermission", "CheckResult")
    private fun ping() {
        Observable.just(Unit)
                .observeOn(Schedulers.io())
                .map { NetworkUtils.isAvailableByPing() }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { isAvailable ->
                    networkTest = isAvailable
                    if (!isAvailable) {
                        onError.onNext("无网络连接")
                    }
                }
    }

    companion object {

        const val NO_SELECT = Int.MIN_VALUE

        private fun getPoiPoint(item: PoiItem): Point {

            val point =
                    if (item.enter != null)
                        item.enter
                    else
                        if (item.exit != null)
                            item.exit
                        else
                            item.latLonPoint
            return Point.box(point)
        }

        private fun getGuideInfo(naviPath: NaviPath): List<GuideInfo> {
            val steps = ArrayList<GuideInfo>()
            val aMapNaviGuides = naviPath.guideList
            val path = naviPath.amapNaviPath
            val aMapNaviSteps = path.steps

            for (j in aMapNaviGuides.indices) {
                val g = aMapNaviGuides[j]
                val group = GuideInfo()
                group.groupIconType = g.iconType
                group.groupLen = g.length
                group.groupName = g.name
                group.groupToll = g.toll
                val count = g.segCount
                val startSeg = g.startSegId
                var traffics = 0
                for (i in startSeg until count + startSeg) {
                    val step = aMapNaviSteps[i]
                    traffics += step.trafficLightNumber
                    var roadName: String
                    if (i == count + startSeg - 1 && j == aMapNaviGuides.size - 1) {
                        roadName = "终点"
                    } else if (i == count + startSeg - 1 && j + 1 < aMapNaviGuides.size - 1) {
                        val ag = aMapNaviGuides[j + 1]
                        roadName = ag.name
                    } else {
                        roadName = step.links[0].roadName
                    }

                    val lbsGuidStep = GuideInfo.Step(step.iconType,
                            roadName, step.length)
                    group.steps.add(lbsGuidStep)

                }
                group.groupTrafficLights = traffics
                steps.add(group)
            }
            return steps
        }

        private fun getPathTime(time: Int): String {
            if (time.toLong() > 3600) {
                val hour = time.toLong() / 3600
                val miniate = time.toLong() % 3600 / 60
                return hour.toString() + "小时" + miniate + "分钟"
            }
            if (time.toLong() >= 60) {
                val miniate = time.toLong() / 60
                return miniate.toString() + "分钟"
            }
            return time.toLong().toString() + "秒"
        }

        private fun getPathLength(path: Int): String {
            if (path > 10000)
            // 10 km
            {
                val dis = (path / 1000).toFloat()
                return dis.toString() + "千米"
            }
            if (path > 1000) {
                val dis = path.toFloat() / 1000
                val fnum = DecimalFormat("##0.0")
                val dstr = fnum.format(dis.toDouble())
                return dstr + "千米"
            }
            if (path > 100) {
                val dis = (path / 50 * 50).toFloat()
                return dis.toString() + "米"
            }
            var dis = (path / 10 * 10).toFloat()
            if (dis == 0f) {
                dis = 10f
            }
            return dis.toString() + "米"
        }
    }
}